### PR TITLE
Fixed resource allocation.

### PR DESCRIPTION
**Pull request by dpoltavets**
Now, when calculating the free resource, we take as a starting value a random value from the specified range. If we could not take a free resource, then we use the minimum value as the initial value.

Closes: #3047.

### CODE REVIEW

**Comment by rtretyak** [comment]
Please un-ignore the `MultiRerouteSpec."Simultaneous reroute of multiple flows should not oversubscribe any ISLs"` functional test. It is able to test the fix

File: src-java/base-topology/base-storm-topology/src/test/java/org/openkilda/wfm/share/flow/resources/MeterPoolTest.java [hunk 1]
```diff
@@ -56,26 +58,11 @@ public void setUp() {
...
-        long fifth = meterPool.allocate(SWITCH_ID, "flow_5", new PathId("path_5")).getValue();
-        assertEquals(33, fifth);
-
-        long sixth = meterPool.allocate(SWITCH_ID, "flow_6", new PathId("path_6")).getValue();
-        assertEquals(34, sixth);
+        Set<MeterId> cookies = new HashSet<>();
+        for (int i = 31; i < 41; i++) {
+            cookies.add(meterPool.allocate(SWITCH_ID, format("flow_%d", i), new PathId(format("path_%d", i))));
+        }
+        assertEquals(10, cookies.size());
```
Comment (niksv) [inline src-java/base-topology/base-storm-topology/src/test/java/org/openkilda/wfm/share/flow/resources/MeterPoolTest.java]: you removed `deallocate()` call.

File: src-java/base-topology/base-storm-topology/src/test/java/org/openkilda/wfm/share/flow/resources/transitvlan/TransitVlanPoolTest.java [hunk 2]
```diff
@@ -49,34 +48,14 @@ public void setUp() {
...
-                .getTransitVlan().getVlan();
-        assertEquals(103, fifth);
+    public void vlanPoolTest() {
+        for (int i = 100; i <= 110; i++) {
+            Flow flow = Flow.builder().flowId(format("flow_%d", i)).srcSwitch(switch1).destSwitch(switch2).build();
+            assertTrue(transitVlanPool.allocate(
+                    flow,
+                    new PathId(format("path_%d", i)),
+                    new PathId(format("opposite_dummy_%d", i))).getTransitVlan().getVlan() > 0);
+        }
```
Comment (niksv) [inline src-java/base-topology/base-storm-topology/src/test/java/org/openkilda/wfm/share/flow/resources/transitvlan/TransitVlanPoolTest.java]: you removed deallocate too

File: src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/flow/resources/ResourcePool.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    protected int computeStartValue(int minValue, int maxValue) {
+        return new Random().ints(1, minValue, maxValue).iterator().nextInt();
+    }
+
+    /**
+     * Return a random long value between minValue and maxValue.
+     */
+    protected long computeStartValue(long minValue, long maxValue) {
+        return new Random().longs(1, minValue, maxValue).iterator().nextLong();
+    }
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/flow/resources/ResourcePool.java]: Fixed.

File: src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/repositories/FlowCookieRepository.java [hunk 4]
```diff
@@ -26,8 +26,9 @@
      * Find an unmasked cookie which is not assigned to any flow.
      * Use the provided {@code defaultCookie} as the first candidate.
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/repositories/FlowCookieRepository.java]: Fixed.

File: src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/repositories/FlowMeterRepository.java [hunk 5]
```diff
@@ -40,8 +40,9 @@
      * Use the provided {@code defaultMeterId} as the first candidate.
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/repositories/FlowMeterRepository.java]: Fixed.

File: src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/repositories/TransitVlanRepository.java [hunk 6]
```diff
@@ -30,8 +30,9 @@
      * Find a transit vlan which is not assigned to any flow.
      * Use the provided defaultVlan as the first candidate.
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/repositories/TransitVlanRepository.java]: Fixed.

File: src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/repositories/VxlanRepository.java [hunk 7]
```diff
@@ -28,8 +28,9 @@
      * Find a vxlan which is not assigned to any flow.
      * Use the provided defaultVxlan as the first candidate.
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/repositories/VxlanRepository.java]: Fixed.

File: src-java/kilda-persistence-neo4j/src/main/java/org/openkilda/persistence/repositories/impl/Neo4jFlowCookieRepository.java [hunk 8]
```diff
@@ -52,22 +52,23 @@ public Neo4jFlowCookieRepository(Neo4jSessionFactory sessionFactory, Transaction
...
 
     @Override
-    public Optional<Long> findUnassignedCookie(long defaultCookie) {
+    public Optional<Long> findUnassignedCookie(long minCookie, long maxCookie) {
         Map<String, Object> parameters = ImmutableMap.of(
-                "default_cookie", defaultCookie);
+                "min_cookie", minCookie,
+                "max_cookie", maxCookie);
 
         // The query returns the default_cookie if it's not used in any flow_cookie,
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/kilda-persistence-neo4j/src/main/java/org/openkilda/persistence/repositories/impl/Neo4jFlowCookieRepository.java]: Fixed.

File: src-java/kilda-persistence-neo4j/src/main/java/org/openkilda/persistence/repositories/impl/Neo4jFlowMeterRepository.java [hunk 9]
```diff
@@ -77,24 +77,25 @@ public Neo4jFlowMeterRepository(Neo4jSessionFactory sessionFactory, TransactionM
...
-    public Optional<MeterId> findUnassignedMeterId(SwitchId switchId, MeterId defaultMeterId) {
+    public Optional<MeterId> findUnassignedMeterId(SwitchId switchId, MeterId minMeterId, MeterId maxMeterId) {
         Map<String, Object> parameters = ImmutableMap.of(
-                "default_meter", defaultMeterId.getValue(),
+                "min_meter", minMeterId.getValue(),
+                "max_meter", maxMeterId.getValue(),
                 "switch_id", switchId.toString()
         );
 
         // The query returns the default_meter if it's not used in any flow_meter,
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/kilda-persistence-neo4j/src/main/java/org/openkilda/persistence/repositories/impl/Neo4jFlowMeterRepository.java]: Fixed.

File: src-java/kilda-persistence-neo4j/src/main/java/org/openkilda/persistence/repositories/impl/Neo4jTransitVlanRepository.java [hunk 10]
```diff
@@ -77,22 +77,23 @@ public Neo4jTransitVlanRepository(Neo4jSessionFactory sessionFactory, Transactio
...
 
     @Override
-    public Optional<Integer> findUnassignedTransitVlan(int defaultVlan) {
+    public Optional<Integer> findUnassignedTransitVlan(int minVlan, int maxVlan) {
         Map<String, Object> parameters = ImmutableMap.of(
-                "default_vlan", defaultVlan);
+                "min_vlan", minVlan,
+                "max_vlan", maxVlan);
 
         // The query returns the default_vlan if it's not used in any transit_vlan,
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/kilda-persistence-neo4j/src/main/java/org/openkilda/persistence/repositories/impl/Neo4jTransitVlanRepository.java]: Fixed.

File: src-java/kilda-persistence-neo4j/src/main/java/org/openkilda/persistence/repositories/impl/Neo4jVxlanRepository.java [hunk 11]
```diff
@@ -50,22 +50,23 @@ public Neo4jVxlanRepository(Neo4jSessionFactory sessionFactory, TransactionManag
...
 
     @Override
-    public Optional<Integer> findUnassignedVxlan(int defaultVni) {
+    public Optional<Integer> findUnassignedVxlan(int minVni, int maxVni) {
         Map<String, Object> parameters = ImmutableMap.of(
-                "default_vni", defaultVni);
+                "min_vni", minVni,
+                "max_vni", maxVni);
 
         // The query returns the default_vni if it's not used in any vxlan,
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/kilda-persistence-neo4j/src/main/java/org/openkilda/persistence/repositories/impl/Neo4jVxlanRepository.java]: Fixed.

File: src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/flow/resources/ResourcePool.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sergii-iakovenko) [reply src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/flow/resources/ResourcePool.java]: I believe this was a very controversial decision. Why should we expose internal pool logic as a dedicated class with static methods? To allow other classes to reuse them? But what for, if they're pool specific? I don't see any benefits in the solution with static methods vs inherited.

There's an option with moving the methods out of the class, create a strategy, and allow pool to choose among implementations.

File: src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/flow/resources/ResourceUtils.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.openkilda.wfm.share.flow.resources;
+
+import java.util.Random;
+
+public final class ResourceUtils {
```
Comment (sergii-iakovenko) [inline src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/flow/resources/ResourceUtils.java]: Please, check my comment about the class.

File: src-java/base-topology/base-storm-topology/src/test/java/org/openkilda/wfm/share/flow/resources/CookiePoolTest.java [hunk 13]
```diff
@@ -35,26 +39,16 @@ public void setUp() {
...
-
-        cookiePool.deallocate(second);
-        long fourth = cookiePool.allocate("flow_4");
-        assertEquals(6, fourth);
-
-        long fifth = cookiePool.allocate("flow_5");
-        assertEquals(8, fifth);
+        Set<Long> cookies = new HashSet<>();
+        for (int i = 5; i <= 25; i++) {
+            cookies.add(cookiePool.allocate(format("flow_%d", i)));
```
Comment (sergii-iakovenko) [inline src-java/base-topology/base-storm-topology/src/test/java/org/openkilda/wfm/share/flow/resources/CookiePoolTest.java]: The test became shallow: it doesn't verify the allocated values and doesn't check deallocation. I think additional tests should be added to cover this.

Comment (dpoltavets) [reply src-java/base-topology/base-storm-topology/src/test/java/org/openkilda/wfm/share/flow/resources/CookiePoolTest.java]: Tests have been added.

**Pull request merged**